### ISSUE TITLE
minetest.clear_craft unbearable slow when using the recipe as "identifier"

### ISSUE THREAD

**Issue by dsohler** [minetest/minetest#5790]
In my mod I invoke `minetest.clear_craft()` a lot (and I mean A LOT) of times automatically because I programmatically replace the crafting result.

When using `minetest.clear_craft({ output = id })` where `id` is the node ID of the node I want to remove the recipe for it all works fine. But when using `minetest.clear_craft({recipe = {{id}}})` (`id` is the node ID, too, because there is a shapeless recipe being only the node).

Example without `minetest.clear_craft({recipe = {{id}}})`.

~~~~
elapsed time: 0.00

[xfurniture] Loading "chairs" objects
[xfurniture] Loading "tables" objects
[xfurniture] Loading "window_stuff" objects
[xfurniture] Loading "benches" objects
[xfurniture] Loading "stairs_slabs" objects
[xfurniture] Loading "walls" objects
[xfurniture] Fetching uncraftable objects
[xfurniture] Replacing recipes of 312 uncraftable objects
elapsed time: 0.04
~~~~

When using  `minetest.clear_craft({recipe = {{id}}})` (all I did was removing the `--` in front of the call so Lua will parse it and not ignore it as comment).

~~~~
elapsed time: 0.00

[xfurniture] Loading "chairs" objects
[xfurniture] Loading "tables" objects
[xfurniture] Loading "window_stuff" objects
[xfurniture] Loading "benches" objects
[xfurniture] Loading "stairs_slabs" objects
[xfurniture] Loading "walls" objects
[xfurniture] Fetching uncraftable objects
[xfurniture] Replacing recipes of 312 uncraftable objects
elapsed time: 57.13
~~~~

... when marking it as comment again the elapsed time is down at 0.04 again.

**Comment by erlehmann** [comment]
@Zughy asked me to verify this last night and yes, this is still *extremely* slow. Here is my benchmark code:

```
_start = os.clock()
for i=1,4096,1 do
	minetest.register_craftitem(
		"benchmark_clear_craft:item_" .. tostring(i),
		{
			description = "item_" .. tostring(i)
		}
	)
end
_end = os.clock()
minetest.debug(
	"time to register 4096 craftitems: " .. tostring(_end - _start)
)

_start = os.clock()
for i=1,4096,1 do
	minetest.register_craft({
		output = "benchmark_clear_craft:item_" .. tostring(i),
		type = "shapeless",
		recipe = { "benchmark_clear_craft:item_" .. tostring(i) }
	})
end
_end = os.clock()
minetest.debug(
	"time to register 4096 crafts: " .. tostring(_end - _start)
)

_start = os.clock()
for i=1,2048,1 do
	minetest.clear_craft({
		output = "benchmark_clear_craft:item_" .. tostring(i),
	})
end
_end = os.clock()
minetest.debug(
	"time to clear 2048 crafts based on output: " .. tostring(_end - _start)
)
_start = os.clock()
for i=2049,4096,1 do
	minetest.clear_craft({
		type = "shapeless",
		recipe = { "benchmark_clear_craft:item_" .. tostring(i) }
	})
end
_end = os.clock()
minetest.debug(
	"time to clear 2048 crafts based on recipe: " .. tostring(_end - _start)
)

--[[
2022-05-12 23:22:26: [Main]: time to register 4096 craftitems: 0.066811
2022-05-12 23:22:26: [Main]: time to register 4096 crafts: 0.065608
2022-05-12 23:22:26: [Main]: time to clear 2048 crafts based on output: 0.01529
2022-05-12 23:22:53: [Main]: time to clear 2048 crafts based on recipe: 24.10416
]]--
```

**Comment by appgurueu** [comment]
Minetest is poorly hashing craft recipes. This affects both clear_craft by recipe and getting the output for a given crafting grid. Only grid recipes without groups are properly hashed. For everything else Minetest falls back to "hashing" by item count, which obviously is very poor and ends up turning the search for craft recipes into a linear search.

For shapeless recipes without groups - such as the ones in these benchmarks - simply hashing them by the list of items, lexicographically sorted, should suffice to improve the performance from linear to constant time.

**Comment by sfan5** [comment]
related: #11224

**Comment by appgurueu** [comment]
> related: #11224

(practically almost a duplicate of that issue, except that this one specifically focuses on clear_craft, but the underlying reason is the same)

**Comment by Desour** [comment]
For recipes with groups, hashing the whole grid would require too much memory to store all recipes. Ie. if there are 10 types of cobble, you'd have 10^8=0.1G recipes for the mtg furnace.

One could instead make something like an automaton that finds the possible recipes for a given grid of items. Recipes are just fixed-length (length=9) languages of words(=grids) consisting of items.

**Comment by appgurueu** [comment]
> For recipes with groups, hashing the whole grid would require too much memory to store all recipes. Ie. if there are 10 types of cobble, you'd have 10^8=0.1G recipes for the mtg furnace.
> 
> One could instead make something like an automaton that finds the possible recipes for a given grid of items. Recipes are just fixed-length (length=9) languages of words(=grids) consisting of items. Edit: And for shapeless crafting recipes, one can first sort the items lexicographically, as was already said.

Yes, that's what makes this problem so hard; I doubt your "automata" approach would necessarily have better worst-case complexity because the involved powerset construction again has exponential time. Again, the problem are the groups. Same for shapeless: Lexicographically sorting groups doesn't quite work.

I like your approach of treating this as a string search problem (which it is). I suppose a proper solution indeed lies in that direction.

**Comment by Desour** [comment]
>Yes, that's what makes this problem so hard; I doubt your "automata" approach would necessarily have better worst-case complexity. Remember the goal is to get sublinear here - RegEx matching is (if implemented correctly) linear and not at all better than the current approach. A suffix tree might be what you're looking for, but fails for "groups" (choice). Again, the problem are the groups.

The current approach is linear in the number of recipes matching with the hash of the input (and for recipes with groups, the hash is just the number of non-empty slots).
What I wanted to suggest would be linear in the number of grid slots (=9):
* Create a simple NFA for each recipe, with 10 states, of which one is final, and use item names at the transitions (=> for groups there are multiple transitions from a state).
* Combine these NFAs via the powerset construction to a DFA, and keep extra information at the resulting final states for which recipes they represent.
  (Each state of the DFA encodes which recipes are still possible after having read a prefix. Minimization makes no sense, I think, because we don't want to merge final states.)
  About the size of the DFA: Let the "depth" of a state (ignoring the trap state) be the distance from the initial state (0...10). Due to the semantics of the powerset construction, at depth d>1, there are at most as many states as in depth d-1. At depth 1 (after the root), there are at most as many states as there are items (very rough estimation). Hence, the size is in `O(<num registered items> * 9)`.
* Then when checking shaped recipes, just run this DFA (takes 9 steps). (Then do more fine grained checks (ie. meta), and choose the recipe with the best priority.)

This obviously doesn't work for shapeless recipes.
The DFA would be somewhat like a trie, but not exactly a tree (ie. there can be diamonds). I'm not sure how a suffix tree can be used here (but I don't know much about those anyways).

>Same for shapeless: Lexicographically sorting groups doesn't quite work.

Oh right. The item name order obviously is different from the orders of the items' groups.
(Btw. recipe grids are already sorted.)

For shapeless recipes, one could run all permutations of the input at once: In the first step, try all 9 slots. For each one that does not end in the (implicit) trap state, store a front, that is, a state with its remaining items. Now update the set of fronts 8 more times like this.

As an optimisation, one can do all of this per recipe size.

**Comment by appgurueu** [comment]
Right, so you're going the powerset route rather than the suffix tree route (my idea was to build a suffix tree of recipe strings and then "searching" for a recipe). This has an exponential worst-case complexity. Is there any restriction on craft recipes that prevents this worst case from happening?

**Comment by Desour** [comment]
>This has an exponential worst-case complexity. Is there any restriction on craft recipes that prevents this worst case from happening?

Yes, see the "About the size of the DFA:" section of my comment.
